### DISCUSSION TITLE
no useable address found for interfaces [eth0 en0]

### DISCUSSION THREAD

**Discussion by toni-ainita-net** [grafana/mimir#8773] [Help and support]
#### Describe the bug
I am trying to start Mimir using the demo.yaml described in https://grafana.com/docs/mimir/latest/get-started/ and geting the error "err="no useable address found for interfaces [eth0 en0]\ninvalid ring lifecycler config".

Full error:
ts=2024-07-19T19:24:00.848790877Z caller=main.go:225 level=info msg="Starting application" version="(version=2.13.0, branch=release-2.13, revision=4775ec1568)"
ts=2024-07-19T19:24:00.854274835Z caller=server.go:352 level=info msg="server listening on addresses" http=[::]:9009 grpc=[::]:9095
ts=2024-07-19T19:24:00.860241138Z caller=api.go:155 level=debug msg="api: registering route" methods=GET path=/config auth=false gzip=true
ts=2024-07-19T19:24:00.860413571Z caller=api.go:155 level=debug msg="api: registering route" methods=GET path=/ auth=false gzip=true
ts=2024-07-19T19:24:00.860430052Z caller=api.go:160 level=debug msg="api: registering route" methods=GET prefix=/static/ auth=false gzip=true
ts=2024-07-19T19:24:00.860442545Z caller=api.go:155 level=debug msg="api: registering route" methods=GET path=/debug/fgprof auth=false gzip=true
ts=2024-07-19T19:24:00.860457363Z caller=api.go:155 level=debug msg="api: registering route" methods=GET path=/api/v1/status/buildinfo auth=false gzip=true
ts=2024-07-19T19:24:00.860470718Z caller=api.go:155 level=debug msg="api: registering route" methods=GET path=/api/v1/status/config auth=false gzip=true
ts=2024-07-19T19:24:00.860495184Z caller=api.go:155 level=debug msg="api: registering route" methods=GET path=/api/v1/status/flags auth=false gzip=true
ts=2024-07-19T19:24:00.860745945Z caller=api.go:155 level=debug msg="api: registering route" methods=GET path=/memberlist auth=false gzip=true
ts=2024-07-19T19:24:00.861775706Z caller=storage.go:39 level=warn msg="-ruler-storage.backend=filesystem is for development and testing only; you should switch to an external object store for production use or use a shared filesystem"
ts=2024-07-19T19:24:00.862250988Z caller=netutil.go:85 level=debug msg="looking for addresses" inf="[eth0 en0]" inet6enabled=false
ts=2024-07-19T19:24:00.862355424Z caller=netutil.go:90 level=warn msg="error getting addresses for interface" inf=eth0 err="route ip+net: no such network interface"
ts=2024-07-19T19:24:00.862417891Z caller=netutil.go:90 level=warn msg="error getting addresses for interface" inf=en0 err="route ip+net: no such network interface"
ts=2024-07-19T19:24:00.862424313Z caller=netutil.go:118 level=debug msg="detected IP address after looking up all configured interface names" ipAddr="invalid IP"
ts=2024-07-19T19:24:00.862450292Z caller=memberlist_client.go:435 level=info msg="Using memberlist cluster label and node name" cluster_label= node=grafana-8ece3768
ts=2024-07-19T19:24:00.862720629Z caller=log.go:103 level=error msg="error running application" err="no useable address found for interfaces [eth0 en0]\ninvalid ring lifecycler config\ngithub.com/grafana/mimir/pkg/storegateway.newStoreGateway\n\t/go/src/github.com/grafana/mimir/pkg/storegateway/gateway.go:153\ngithub.com/grafana/mimir/pkg/storegateway.NewStoreGateway\n\t/go/src/github.com/grafana/mimir/pkg/storegateway/gateway.go:122\ngithub.com/grafana/mimir/pkg/mimir.(*Mimir).initStoreGateway\n\t/go/src/github.com/grafana/mimir/pkg/mimir/modules.go:966\ngithub.com/grafana/dskit/modules.(*Manager).initModule\n\t/go/src/github.com/grafana/mimir/vendor/github.com/grafana/dskit/modules/modules.go:136\ngithub.com/grafana/dskit/modules.(*Manager).InitModuleServices\n\t/go/src/github.com/grafana/mimir/vendor/github.com/grafana/dskit/modules/modules.go:108\ngithub.com/grafana/mimir/pkg/mimir.(*Mimir).Run\n\t/go/src/github.com/grafana/mimir/pkg/mimir/mimir.go:852\nmain.main\n\t/go/src/github.com/grafana/mimir/cmd/mimir/main.go:227\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:271\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1695\nerror initialising module: store-gateway\ngithub.com/grafana/dskit/modules.(*Manager).initModule\n\t/go/src/github.com/grafana/mimir/vendor/github.com/grafana/dskit/modules/modules.go:138\ngithub.com/grafana/dskit/modules.(*Manager).InitModuleServices\n\t/go/src/github.com/grafana/mimir/vendor/github.com/grafana/dskit/modules/modules.go:108\ngithub.com/grafana/mimir/pkg/mimir.(*Mimir).Run\n\t/go/src/github.com/grafana/mimir/pkg/mimir/mimir.go:852\nmain.main\n\t/go/src/github.com/grafana/mimir/cmd/mimir/main.go:227\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:271\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1695"

#### To Reproduce
./mimir  --config.file=./demo.yaml 

1. Start Mimir (SHA or version)
2. Perform Operations(Read/Write/Others)

#### Expected behavior

Mimir should start.

#### Environment

Mimir version (2.13.0)
Debian 12
Deployment method (binary)

**Comment by valentino-prt** [comment]
Same problem here using v2.15.1, someone found a solution?

**Comment by friedow** [comment]
In the default configuration Mimir only looks on the `eth0` and `en0` network interfaces for an address to bind. If your machine does not have those interfaces, mimir will fail with `no useable address found for interfaces [eth0 en0]` which is what you are seing.

To fix this you need to figure out which network interface your machine uses and configure the different mimir components to look for this interface.

To determine which network interfaces you've got ssh into your machine and run `ifconfig -a`. This will list all available network interfaces and their respective ip addresses.

Lets assume the network interface you want to use is `enp1s0`. Now you can configure the different mimir components to use it by adding the following to you mimir config. **Change the interface name below if your network interface is named differently!**
```yaml
distributor:
  ring:
    instance_interface_names:
      - enp1s0
ingester:
  ring:
    instance_interface_names:
      - enp1s0
frontend:
  instance_interface_names:
    - enp1s0
compactor:
  sharding_ring:
    instance_interface_names:
      - enp1s0
store_gateway:
  sharding_ring:
    instance_interface_names:
      - enp1s0
ruler:
  ring:
    instance_interface_names:
      - enp1s0
alertmanager:
  sharding_ring:
    instance_interface_names:
      - enp1s0
query_scheduler:
  ring:
    instance_interface_names:
      - enp1s0
```

Related Links:
- https://wiggels.dev/posts/error-starting-mimir-with-predictable-interface-names/
- https://grafana.com/docs/mimir/latest/configure/configuration-parameters/

**Comment by faxm0dem** [comment]
For a first time user like me, figuring out this problem is really a pain. I had to start / read journald errors / start to find out all subsystems that needed the same exact configuration option.
Using a clean install of 3.1.2 I had to set `instance_interface_names` 7 times or mimir wouldn't start. The mimir [getting started documentation](https://grafana.com/docs/mimir/latest/get-started/) failed to get started as it's only specifying the setting twice.

Hard-coding en0 and eth0 not only seems strange to me, but gives a bad impression on the quality of the software.

**Comment by s3onghyun** [comment]
This comes from the ring lifecycler trying to auto-detect the address to advertise. By default it only looks at interfaces named `eth0` and `en0`, and if your host has neither (common on bare metal or VMs with names like `enp1s0`, and in some container runtimes) it fails with exactly this `no useable address found for interfaces [eth0 en0]`, followed by `invalid ring lifecycler config`.

Find your real interface (`ip addr` or `ifconfig -a`) and tell Mimir about it. On current Mimir you can set it once in the common config and it propagates to every ring:

```yaml
common:
  instance_interface_names: [enp1s0]
```

If you are on an older config layout, `instance_interface_names` has to be set under each ring (ingester, distributor, compactor, store_gateway, ruler, and so on), which is why the getting-started demo breaks on non-standard hosts. Two alternatives that also skip interface discovery entirely: set the address directly with `instance_addr`, or set `-memberlist.advertise-addr`.
